crypto/src/main/java/org/web3j/crypto/StructuredDataEncoder.java

```diff
@@ -0,0 +1,448 @@
+        HashSet<String> deps = new HashSet<>();
+
+        if (!types.containsKey(primaryType)) {
+            return deps;
+        }
+
+        List<String> remainingTypes = new ArrayList<>();
+        remainingTypes.add(primaryType);
+
+        while (remainingTypes.size() > 0) {
+            String structName = remainingTypes.get(remainingTypes.size() - 1);
+            remainingTypes.remove(remainingTypes.size() - 1);
+            deps.add(structName);
+
+            Iterator itr = types.get(primaryType).iterator();
+            while (itr.hasNext()) {
+                StructuredData.Entry entry = (StructuredData.Entry) itr.next();
+                if (!types.containsKey(entry.getType())) {
+                    // Don't expand on non-user defined types
+                    continue;
+                } else if (deps.contains(entry.getType())) {
+                    // Skip types which are already expanded
+                    continue;
+                } else {
+                    // Encountered a user defined type
+                    remainingTypes.add(entry.getType());
+                }
+            }
+        }
+
+        return deps;
+    }
+
+    public String encodeStruct(String structName) {
+        HashMap<String, List<StructuredData.Entry>> types = jsonMessageObject.getTypes();
+
+        String structRepresentation = structName + "(";
+        for (StructuredData.Entry entry: types.get(structName)) {
+            structRepresentation += String.format("%s %s,", entry.getType(), entry.getName());
+        }
+        structRepresentation = structRepresentation.substring(0, structRepresentation.length() - 1);
+        structRepresentation += ")";
+
+        return structRepresentation;
+    }
+
+    public String encodeType(String primaryType) {
+        HashSet<String> deps = getDependencies(primaryType);
+        deps.remove(primaryType);
+
+        // Sort the other dependencies based on Alphabetical Order and finally add the primaryType
+        List<String> depsAsList = new ArrayList<>(deps);
+        Collections.sort(depsAsList);
+        depsAsList.add(0, primaryType);
+
+        String result = "";
+        for (String structName: depsAsList) {
+            result += encodeStruct(structName);
+        }
+
+        return result;
+    }
+
+    public byte[] typeHash(String primaryType) {
+        return Numeric.hexStringToByteArray(sha3String(encodeType(primaryType)));
+    }
+
+    public List<Integer> getArrayDimensionsFromDeclaration(String declaration) {
+        Pattern arrayTypePattern = Pattern.compile(arrayTypeRegex);
+
+        // This regex tries to extract the dimensions from the
+        // square brackets of an array declaration using the ``Regex Groups``.
+        // Eg- It extracts ``5, 6, 7`` from ``[5][6][7]``
+        String arrayDimensionRegex = "\\[([1-9]\\d*)?\\]";
+        Pattern arrayDimensionPattern = Pattern.compile(arrayDimensionRegex);
```

Comment:
Could you make this a field, too? No point compiling the (static) pattern each time the function is called.